tests/t0122_gateway_tar_test.go

```diff
@@ -69,16 +68,19 @@ func TestGatewayTar(t *testing.T) {
 					Header("Content-Disposition").Contains("attachment;"),
 					Header("Etag").Contains("W/\"%s.x-tar", fileCID),
 					Header("Content-Type").Contains("application/x-tar"),
-				),
-				// TODO: try to compare bytes for bytes.
-				// .Body(
-				// Tar().
-				// 	IsValid().
-				// 	File("outputDir/$FILE_CID").Content(
-				// 	Equals("I am a txt file on path with utf8"),
-				// ),
-			// TODO: `tar -x` the body.
-			// TODO: test the files and contents.
+				).Body(
+				Shell().
+					WithBody("OUTPUT_FILE").
+					With("FILE_CID", fileCID).
+					Succeeds(`
+						cat $OUTPUT_FILE | tar -x;
+						mkdir outputDir;
+						cat $OUTPUT_FILE | tar -x -C outputDir;
+						echo "I am a txt file on path with utf8" > expected
+						test -f outputDir/$FILE_CID;
+						diff -u expected outputDir/$FILE_CID
+						`),
```

Comment:
Could we try to stay Go native? I feel this introduces too big of an opportunity for error. For example, with shell, it's really hard to handle cleanup properly. 

Deferring to shell also ties us to specific system requirements which would be nice to avoid if possible.

It seems to me that with introduction of shell-based checks, we'd get dangerously close to just hiding sharness underneath Go Tests. And, with all the amazing work you've already done with tests, I think that'd be a step backwards.

I'd be happy to try to create a proposal for how we could handle tars "natively". Just to see if that's at all possible, I put together a Go tar reader which seems to enable exactly this - https://gist.github.com/galargh/7a1baec0f679605e0ba3ecc0d7e558fc

Reply:
Agreed, I don't want the shell to become a new default in the tests either, but I want to be pragmatic about it: we can use some to help us bridge the gap between the sharness and this test suite.

After all, we are migrating a shell-based test suite to Go, so we're not adding "dirty" shell code, but rather removing it.

Each PR has to accomplish a lot:

- The tedious work of porting the tests from sharness to the test suite.
- Some refactoring of the tests to make them nicer and less awkward (like merging cases when we test multiple headers on one request),
- And then implementing new APIs.

Implementing a new API, especially for tar, is not very hard, but it does introduce scope creep in each PR, and we have to anticipate the use cases in advance, even for a single small test.

I would like to use shell to avoid solving three problems with each PR. Once we have a complete test suite, we can go back and make native what needs to be native.

As for the specific issue of cleanup, the checker does take care of it.

Reply:
I don't agree that tar handling is a case that warrants adding a whole new, extremely powerful API which interacts with shell directly.

If we don't want to work on adding new APIs at the same time as porting tests, we could skip adding tests that cannot be handled with the current APIs and document them instead. That would give us an overview of what APIs are missing at the end of the process.

To reduce the number of things we're doing in the test port PRs, I think we could also skip the consolidation for now.

Reply:
> I'm worried we spend time implementing one-time use APIs instead of working on critical stuff like IPNS and implementing spec-related tests & features.

If it's taking us away from more critical work, then I think we shouldn't be spending time on trying to handle those test cases right now at all.

Reply:
Tackling fewer input tests per PR might also help reduce scope.

Reply:
One more thought - the tar handling doesn't really have to become API straight away. It could be a custom inline check  too. Something like:
```
Response().Body().Check(func (body byte[]) error {
  // check if the body is the tar we expect
})
```

That would remove the need for trying to guess the interface in advance while not having to resort to interacting with shell.

Reply:
I created a PR in https://github.com/ipfs/gateway-conformance/pull/28 to review 122 specifically.

> If we don't want to work on adding new APIs at the same time as porting tests, we could skip adding tests that cannot be handled with the current APIs and document them instead. That would give us an overview of what APIs are missing at the end of the process.

It really depends on our goal, between tradeoffs I'd pick the one where we get more tests from more contributors, so:

between:
- spending time implementing and maintaining 50 lines of Go code that will only be used once, and redoing it for every missing API,
- or leaving todo comments in the code,

We benefit more from having a test that runs thanks to a few lines of shell script. It's an intermediate solution that we can reuse when two lines of bash do the job.

And that's true for external contributors as well: right now, if you want to test a behavior not covered by the API, even just to create a repro and propose a PR, you have to be pretty familiar with golang and the PLABs stack.